tests/test_freqtradebot.py

```diff
@@ -4698,3 +4698,95 @@ def test_update_funding_fees(mocker, default_conf, trading_mode, calls, time_mac
+        funding_fee = nominal_value * funding_rate
+        contract_size = 123
+        "LTC/BTC"
+            time: 0, mark: 3.3, fundRate: 0.00032583, nominal_value: 405.9, fundFee: 0.132254397
+            time: 8, mark: 3.2, fundRate: 0.00024472, nominal_value: 393.6, fundFee: 0.096321792
+        "ETH/BTC"
+            time: 0, mark: 2.4, fundRate: 0.0001, nominal_value: 295.2, fundFee: 0.02952
+            time: 8, mark: 2.5, fundRate: 0.0001, nominal_value: 307.5, fundFee: 0.03075
+        "ETC/BTC"
+            time: 0, mark: 4.3, fundRate: 0.00031077, nominal_value: 528.9, fundFee: 0.164366253
+            time: 8, mark: 4.1, fundRate: 0.00022655, nominal_value: 504.3, fundFee: 0.114249165
+        "XRP/BTC"
+            time: 0, mark: 1.2, fundRate: 0.00049426, nominal_value: 147.6, fundFee: 0.072952776
+            time: 8, mark: 1.2, fundRate: 0.00032715, nominal_value: 147.6, fundFee: 0.04828734
+    '''
+    time_machine.move_to("2021-09-01 00:00:00")
+
+    funding_rates = {
+        "LTC/BTC": {
+            1630454400000: 0.00032583,
+            1630483200000: 0.00024472,
+        },
+        "ETH/BTC": {
+            1630454400000: 0.0001,
+            1630483200000: 0.0001,
+        },
+        "ETC/BTC": {
+            1630454400000: 0.00031077,
+            1630483200000: 0.00022655,
+        },
+        "XRP/BTC": {
+            1630454400000: 0.00049426,
+            1630483200000: 0.00032715,
+        }
+    }
+
+    mark_prices = {
+        "LTC/BTC": {
+            1630454400000: 3.3,
+            1630483200000: 3.2,
+        },
+        "ETH/BTC": {
+            1630454400000: 2.4,
+            1630483200000: 2.5,
+        },
+        "ETC/BTC": {
+            1630454400000: 4.3,
+            1630483200000: 4.1,
+        },
+        "XRP/BTC": {
+            1630454400000: 1.2,
+            1630483200000: 1.2,
+        }
+    }
+
+    mocker.patch(
+        'freqtrade.exchange.Exchange._get_mark_price_history',
+        side_effect=[
+            mark_prices["LTC/BTC"],
+            mark_prices["ETH/BTC"],
+            mark_prices["ETC/BTC"],
+            mark_prices["XRP/BTC"],
+        ]
+    )
+    mocker.patch(
+        'freqtrade.exchange.Exchange.get_funding_rate_history',
+        side_effect=[
+            funding_rates["LTC/BTC"],
+            funding_rates["ETH/BTC"],
+            funding_rates["ETC/BTC"],
+            funding_rates["XRP/BTC"],
+        ]
+    )
+    patch_RPCManager(mocker)
+    patch_exchange(mocker)
+    default_conf['trading_mode'] = 'futures'
+    default_conf['collateral'] = 'isolated'
+    default_conf['dry_run'] = True
+    freqtrade = get_patched_freqtradebot(mocker, default_conf)
+    create_mock_trades(fee, False)
+    time_machine.move_to("2021-09-01 08:00:00 +00:00")
+    freqtrade._schedule.run_pending()
+
+    trades = Trade.get_open_trades()
+    for trade in trades:
+        assert trade.funding_fees == 123 * mark_prices[trade.pair] * funding_rates[trade.pair]
```

Comment:
@xmatthias @rokups  When I run `get_open_trades`, I get these 4 `LTC/BTC`, `ETH/BTC`, `ETC/BTC`, `XRP/BTC`, so I'm trying to test that the funding fees are what they are supposed to be after 8 hours

Currently I'm getting the error below, do you understand what I have to do for this trade to work

```
self = <MagicMock name='get_funding_rate_history' id='4900146912'>
args = ('ETH/BTC', 1630496580283), kwargs = {}
effect = <list_iterator object at 0x12412bdc0>

    def _execute_mock_call(self, /, *args, **kwargs):
        # separate from _increment_mock_call so that awaited functions are
        # executed separately from their call, also AsyncMock overrides this method
    
        effect = self.side_effect
        if effect is not None:
            if _is_exception(effect):
                raise effect
            elif not _callable(effect):
>               result = next(effect)
E               StopIteration

/usr/local/Cellar/python@3.9/3.9.7_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/unittest/mock.py:1154: StopIteration
```

Reply:
I think you call `get_funding_rate_history()` more times than there are mocked values provided

Reply:
That's correct - whenever you mock with `side_effect=[` (side-effect is an array) - that array must be at least as long as the amount of calls you're going to make.
That doesn't seem to be the case here, at least that's what the error suggests.

I'd probably not use `side_effect` this way here - but rather write a (test-local) method returning the correct value based on the pair passed in (so you're not relying on the order of the pairs - which i don't think is guaranteed the way we query trades - but is not a problem either).
We're doing this elsewhere too (iirc in backtesting tests) - which might serve as an example.

Reply:
> test-local) method returning the correct value based on the pair passed in 

What does that look like

Reply:
I changed side_effect to lambdas. There is still an issue with assert as it tries to multiply dicts. It probably should be this:
> assert trade.funding_fees == 123 * mark_prices[trade.pair][arrow.utcnow().int_timestamp * 1000] * funding_rates[trade.pair][arrow.utcnow().int_timestamp * 1000]
But im not entirely sure so i did leave it out. Even if this bit is correct assert still fails.

Reply:
thanks